### PR TITLE
Miscellanous refactorings

### PR DESCRIPTION
**Pull request by mangstadt**
Please only pay attention to the commits on May 5. I don't know why it's including my entire commit history, sorry! :(. They contain minor refactorings to improve performance and readability.

### CODE REVIEW

**Comment by PawelAdamski** [comment]
One more time thanks for your contribution. I will try to check it by the end of this week.

**Comment by mangstadt** [comment]
No problem. If it's too confusing to review/merge with all the old commits, please let me know and I will resubmit it.

File: src/main/java/com/github/paweladamski/httpclientmock/HttpClientMock.java [hunk 1]
```diff
@@ -108,16 +114,16 @@ public HttpClientMockBuilder onDelete() {
...
 
   /**
-   * Starts defining new rule which requires HTTP OPTION method.
+   * Starts defining new rule which requires HTTP OPTIONS method.
    *
    * @return HttpClientMockBuilder which allows  to define new rule
    */
-  public HttpClientMockBuilder onOption() {
-    return newRule("OPTION");
+  public HttpClientMockBuilder onOptions() {
```
Comment (PawelAdamski) [inline src/main/java/com/github/paweladamski/httpclientmock/HttpClientMock.java:136]: You are right it should be `onOptions`. Nevertheless, I would keep method `onOption` for backward compatibility and mark it as @Depracted.

File: src/main/java/com/github/paweladamski/httpclientmock/HttpClientVerifyBuilder.java [hunk 2]
```diff
@@ -216,7 +216,7 @@ public void called(Matcher<Integer> numberOfCalls) {
         .count();
 
     if (!numberOfCalls.matches(matchingCalls)) {
-      throw new IllegalStateException(String.format("Expected %s calls, but found %s.", numberOfCalls, matchingCalls));
+      throw new AssertionError(String.format("Expected %s calls, but found %s.", numberOfCalls, matchingCalls));
```
Comment (PawelAdamski) [inline src/main/java/com/github/paweladamski/httpclientmock/HttpClientVerifyBuilder.java]: I would not use `AssertionError` it. It is supposed to be used in a different context.  Please stay with IllegalStateException or define something like HttpClientMockException.

File: src/main/java/com/github/paweladamski/httpclientmock/matchers/HttpResponseMatchers.java [hunk 3]
```diff
@@ -70,14 +66,14 @@ public void describeTo(Description description) {
...
-          }
+        if (cookieStore == null) {
+          return null;
         }
-        return null;
+
+        Optional<Cookie> cookie = cookieStore.getCookies().stream()
+          .filter(c -> c.getName().equalsIgnoreCase(cookieName))
+        .findFirst();
+        return cookie.isPresent() ? cookie.get().getValue() : null;
```
Comment (PawelAdamski) [inline src/main/java/com/github/paweladamski/httpclientmock/matchers/HttpResponseMatchers.java]: You can write:
```
return cookieStore.getCookies().stream()
          .filter(c -> c.getName().equalsIgnoreCase(cookieName))
        .findFirst()
        .map(cookie->cookie.get().getValue())
        .orElse(null);

File: src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java [hunk 4]
```diff
@@ -1,333 +1,333 @@
...
-   * Adds action which returns provided URL-encoded parameter response with status 200. Additionally it sets "Content-type" header to "application/x-www-form-urlencoded".
-   *
-   * @param parameters parameters to return
-   * @return response builder
-   */
-  public HttpClientResponseBuilder doReturnFormParams(Collection<NameValuePair> parameters, Charset charset) {
-    return responseBuilder.doReturnFormParams(parameters, charset);
-  }
-}
+package com.github.paweladamski.httpclientmock;
```
Comment (PawelAdamski) [inline src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java:1]: Can you tell me what has been changed in this file?

File: src/test/java/com/github/paweladamski/httpclientmock/HttpClientResponseBuilderTest.java [hunk 5]
```diff
@@ -286,15 +286,12 @@ public void should_not_throw_exception_when_body_matcher_is_present_on_post_requ
     HttpResponse response = httpClientMock.execute(httpGet("http://localhost:8080/path2"));
     Assert.assertThat(response, hasStatus(SC_NOT_FOUND));
   }
-
+  
   @Test
-  public void doReturnFormParams_should_returnResponseEntityWithFormParameters() throws IOException {
+  public void doReturnFormParams() throws IOException {
```
Comment (PawelAdamski) [inline src/test/java/com/github/paweladamski/httpclientmock/HttpClientResponseBuilderTest.java]: Please don't change this test name and body.

File: src/test/java/com/github/paweladamski/httpclientmock/HttpClientVerifyTest.java [hunk 6]
```diff
@@ -1,288 +1,288 @@
...
+        .notCalled();
+    httpClientMock.verify()
+        .post("http://localhost#xyz")
+        .called(1);
+    httpClientMock.verify()
+        .post("http://localhost")
+        .notCalled();
+  }
+
+  @Test(expected = AssertionError.class)
```
Comment (PawelAdamski) [inline src/test/java/com/github/paweladamski/httpclientmock/HttpClientVerifyTest.java]: As I explained in previous comment AssertionError should not be used here.

**Review by PawelAdamski — CHANGES_REQUESTED** [review]

File: src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java (hunk 4 above)
Comment (mangstadt) [reply src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java:1]: Some of the information in the Javadocs was wrong with respect to character encoding, and some of the wording was inconsistent.

Line 285
Old: `Adds action which returns provided JSON in provided encoding and status 200. Additionally it sets "Content-type" header to "application/json".`
New: `Adds action which returns provided JSON in provided charset and status 200. Additionally it sets "Content-type" header to "application/json".`

Line 305
Old: `Adds action which returns provided XML in UTF-8 and status 200. Additionally it sets "Content-type" header to "application/xml".`
New: `Adds action which returns provided XML in provided charset and status 200. Additionally it sets "Content-type" header to "application/xml".`

Line 325
Old: `Adds action which returns provided URL-encoded parameter response with status 200. Additionally it sets "Content-type" header to "application/x-www-form-urlencoded".`
New: `Adds action which returns provided URL-encoded parameter response in provided charset and status 200. Additionally it sets "Content-type" header to "application/x-www-form-urlencoded".`

File: src/test/java/com/github/paweladamski/httpclientmock/HttpClientResponseBuilderTest.java (hunk 5 above)
Comment (mangstadt) [reply src/test/java/com/github/paweladamski/httpclientmock/HttpClientResponseBuilderTest.java]: I don't know what happened here. Maybe I should re-submit my pull request.

File: src/main/java/com/github/paweladamski/httpclientmock/HttpClientMock.java (hunk 1 above)
Comment (mangstadt) [reply src/main/java/com/github/paweladamski/httpclientmock/HttpClientMock.java:136]: Fixed in 0769442c9cb6357884e04b4875d1eebaf56bde45

File: src/main/java/com/github/paweladamski/httpclientmock/HttpClientVerifyBuilder.java (hunk 2 above)
Comment (mangstadt) [reply src/main/java/com/github/paweladamski/httpclientmock/HttpClientVerifyBuilder.java]: Ok, changed it back. 7e629fe9aeb8dbbff09615e3b4e0e12c6096fa2c

File: src/main/java/com/github/paweladamski/httpclientmock/matchers/HttpResponseMatchers.java (hunk 3 above)
Comment (mangstadt) [reply src/main/java/com/github/paweladamski/httpclientmock/matchers/HttpResponseMatchers.java]: Fixed in b659b35b08fb18e8e6c362d073ce5021207188c3

File: src/test/java/com/github/paweladamski/httpclientmock/HttpClientVerifyTest.java (hunk 6 above)
Comment (mangstadt) [reply src/test/java/com/github/paweladamski/httpclientmock/HttpClientVerifyTest.java]: Fixed in 7e629fe9aeb8dbbff09615e3b4e0e12c6096fa2c

File: src/test/java/com/github/paweladamski/httpclientmock/HttpClientResponseBuilderTest.java (hunk 5 above)
Comment (mangstadt) [reply src/test/java/com/github/paweladamski/httpclientmock/HttpClientResponseBuilderTest.java]: Fixed in 459f3036b2aedfe6e1d3ee27e10c127729ce52ba. I restored your version of the file.

I think the file got messed up when I attempted to merge your branch into my branch after the 1.7.0 release. Sorry about that!

File: src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java (hunk 4 above)
Comment (PawelAdamski) [reply src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java:1]: I was rather asking why the whole file has changed? Have you changed encoding or line endings?

**Review by PawelAdamski — CHANGES_REQUESTED** [review]
Looks good now, I just would like to understand why some files look like completely changed.

File: src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java (hunk 4 above)
Comment (mangstadt) [reply src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java:1]: My git configuration is a little wonky. I am using Windows and git hasn't been doing the line ending conversions when I clone a repo. So, before I commit anything, I have been manually telling Eclipse to convert the line endings to Unix-style. I may have forgotten to do that with this file.

Comment (mangstadt) [reply src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java:1]: Here is a better view of the changes: https://github.com/PawelAdamski/HttpClientMock/pull/22/commits/9e0dcefa4a150da313990de682af928be3399033

**Review by PawelAdamski — APPROVED** [review]

**Pull request merged**